fleetspeak/src/server/mysql/messagestore.go

```diff
@@ -225,32 +225,121 @@ func genPlaceholders(num int) string {
 	return strings.Join(es, ", ")
 }
 
-func (d *Datastore) DeletePendingMessages(ctx context.Context, ids []common.ClientID) error {
-	return d.runInTx(ctx, false, func(tx *sql.Tx) error {
+func (d *Datastore) getPendingMessageRawIds(ctx context.Context, tx *sql.Tx, ids []common.ClientID, offset uint64, limit uint64, forUpdate bool) ([][]byte, error) {
+	squery := fmt.Sprintf("SELECT "+
+		"m.message_id AS message_id "+
+		"FROM messages AS m, pending_messages AS pm "+
+		"WHERE m.destination_client_id IN (%s) AND m.message_id=pm.message_id "+
+		"ORDER BY message_id",
+		genPlaceholders((len(ids))))
+
+	if limit != 0 {
+		squery += " LIMIT ?"
+	}
+
+	if offset != 0 {
+		squery += " OFFSET ?"
+	}
+
+	if forUpdate {
+		squery += " FOR UPDATE"
+	}
+
+	args := make([]interface{}, len(ids), len(ids)+2)
+	for i, v := range ids {
+		args[i] = v.Bytes()
+	}
+
+	if limit != 0 {
+		args = append(args, limit)
+	}
+
+	if offset != 0 {
+		args = append(args, offset)
+	}
+
+	idsToProc := make([][]byte, 0)
+
+	rs, err := tx.QueryContext(ctx, squery, args...)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to fetch the list of messages to delete: %v", err)
+	}
+	defer rs.Close()
+	for rs.Next() {
+		var id []byte
+		if err := rs.Scan(&id); err != nil {
+			return nil, err
+		}
+		idsToProc = append(idsToProc, id)
+	}
+
+	return idsToProc, nil
+}
+
+func (d *Datastore) GetPendingMessageCount(ctx context.Context, ids []common.ClientID) (uint64, error) {
+	var result uint64
+
+	err := d.runInTx(ctx, true, func(tx *sql.Tx) error {
 		squery := fmt.Sprintf("SELECT "+
-			"m.message_id "+
+			"COUNT(*) "+
 			"FROM messages AS m, pending_messages AS pm "+
-			"WHERE m.destination_client_id IN (%s) AND m.message_id=pm.message_id "+
-			"FOR UPDATE", genPlaceholders((len(ids))))
+			"WHERE m.destination_client_id IN (%s) AND m.message_id=pm.message_id ",
+			genPlaceholders((len(ids))))
+		fmt.Println(squery)
 
 		args := make([]interface{}, len(ids))
 		for i, v := range ids {
 			args[i] = v.Bytes()
 		}
-
-		idsToProc := make([]interface{}, 0)
-
 		rs, err := tx.QueryContext(ctx, squery, args...)
 		if err != nil {
-			return fmt.Errorf("Failed to fetch the list of messages to delete: %v", err)
+			return fmt.Errorf("Failed to fetch the pending message count: %v", err)
 		}
 		defer rs.Close()
-		for rs.Next() {
-			var id []byte
-			if err := rs.Scan(&id); err != nil {
+		if !rs.Next() {
+			return fmt.Errorf("Got empty result")
+		}
+		err = rs.Scan(&result)
+		if err != nil {
+			return fmt.Errorf("Failed to scan result: %v", err)
+		}
+		return nil
+	})
+
+	return result, err
+}
+
+func (d *Datastore) GetPendingMessages(ctx context.Context, ids []common.ClientID, offset uint64, count uint64, wantData bool) ([]*fspb.Message, error) {
+	var res []*fspb.Message
+	err := d.runInTx(ctx, true, func(tx *sql.Tx) error {
+		messageIdsRaw, err := d.getPendingMessageRawIds(ctx, tx, ids, offset, count, false)
```

Comment:
Is there any reason why we have to do the query in multiple steps - first get the ids, then get the messages - instead of doing everything in a single SELECT statement over two tables? (similar to how it's done in the `GetPendingMessagesCount`)?

Reply:
* I wanted to avoid re-implementing the logic already existing in `GetMessages`. I.e. if something changes/evolves in that logic, I wanted it to automatically be applied to `GetPendingMessages` as well.
* I also wanted to reuse the logic from `DeletePendingMessages`, which defines what a pending message is.
* `GetPendingMessagesCount` is an unfortunate outlier here, I had to duplicate the SQL query :-(

Reply:
Thanks for providing the background - sounds good. My only concern here would be performance, but given that this will be a rarely used method, a minor performance win that we could get by using a single SQL statement is likely not worth the additional implementation cost.